// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
edition = "2023";

package koladata.ext.persisted_data;

message PersistedIncrementalDataBagManagerMetadata {
  // The version of the manager. Always set. Example format: "1.0.0"
  string version = 1;

  // The number of times the metadata was updated. Always set. Starts at 0 and
  // is incremented by 1 with each update.
  // This is used to make sure that updates to the manager are transactional.
  // It helps to detect concurrent writers and prevents them from overwriting
  // each other's changes.
  int64 metadata_update_number = 3;

  // The metadata of the data bags that are managed by the manager. The order
  // here reflects the order in which the bags were added to the manager.
  // The initial bag is always at index 0.
  repeated DataBagMetadata data_bag_metadata = 2;
}

message DataBagMetadata {
  // The name of the bag. Always set. Unique within the manager.
  string name = 1;

  // The names of the bags that this bag depends on.
  repeated string dependencies = 2;

  // The filename of the serialized bag. Always set.
  // Example: 'bag-000000000001.kd'. The filename refers to a file that is in
  // the directory specified by the `directory` field below.
  string filename = 3;

  // Optional. The directory in which the bag is stored. If not set, then it is
  // assumed to be the persistence directory of the manager.
  string directory = 4;
}
